pkg/controller/restore_finalizer_controller.go

```diff
@@ -200,14 +234,137 @@ func (r *restoreFinalizerReconciler) finishProcessing(restorePhase velerov1api.R
 // finalizerContext includes all the dependencies required by finalization tasks and
 // a function execute() to orderly implement task logic.
 type finalizerContext struct {
-	log logrus.FieldLogger
+	logger          logrus.FieldLogger
+	restore         *velerov1api.Restore
+	crClient        client.Client
+	volumeInfo      []*internalVolume.VolumeInfo
+	restoredPVCList map[string]struct{}
 }
 
 func (ctx *finalizerContext) execute() (results.Result, results.Result) { //nolint:unparam //temporarily ignore the lint report: result 0 is always nil (unparam)
 	warnings, errs := results.Result{}, results.Result{}
 
 	// implement finalization tasks
-	ctx.log.Debug("Starting running execute()")
+	pdpErrs := ctx.patchDynamicPVWithVolumeInfo()
+	errs.Merge(&pdpErrs)
 
 	return warnings, errs
 }
+
+// patchDynamicPV patches newly dynamically provisioned PV using volume info
+// in order to restore custom settings that would otherwise be lost during dynamic PV recreation.
+func (ctx *finalizerContext) patchDynamicPVWithVolumeInfo() (errs results.Result) {
+	ctx.logger.Info("patching newly dynamically provisioned PV starts")
+
+	var pvWaitGroup sync.WaitGroup
+	var resultLock sync.Mutex
+	for _, volumeItem := range ctx.volumeInfo {
+		if volumeItem.BackupMethod == "PodVolumeBackup" || volumeItem.BackupMethod == "CSISnapshot" ||
+			(volumeItem.BackupMethod == "" && !volumeItem.Skipped && volumeItem.PVInfo.ReclaimPolicy == string(v1.PersistentVolumeReclaimDelete)) {
```

Comment:
Please double check if there's a real use case for this condition.

Reply:
Seems that Velero doesn't necessarily handle this user case where PV's content won't get backed up. 

Will remove this condition.